Realm/RLMUtil.mm

```diff
@@ -136,19 +151,40 @@ bool RLMIsSwiftObjectClass(Class cls) {
 }
 
 BOOL RLMValidateValue(__unsafe_unretained id const value,
-                      RLMPropertyType type, bool optional, bool array,
+                      RLMPropertyType type,
+                      bool optional,
+                      RLMCollectionType collectionType,
                       __unsafe_unretained NSString *const objectClassName) {
     if (optional && !RLMCoerceToNil(value)) {
         return YES;
     }
-    if (array) {
+    if (collectionType == RLMCollectionTypeArray) {
         if (auto rlmArray = asRLMArray(value)) {
             return checkArrayType(rlmArray, type, optional, objectClassName);
         }
         if (id enumeration = RLMAsFastEnumeration(value)) {
             // check each element for compliance
             for (id el in enumeration) {
-                if (!RLMValidateValue(el, type, optional, false, objectClassName)) {
+                if (!RLMValidateValue(el, type, optional, RLMCollectionTypeNone, objectClassName)) {
+                    return NO;
+                }
+            }
+            return YES;
+        }
+        if (!value || value == NSNull.null) {
+            return YES;
+        }
+        return NO;
+    }
+
+    if (collectionType == RLMCollectionTypeSet) {
```

Comment:
There doesn't appear to be any reason to distinguish arrays and sets here?

Reply:
I've also removed `RLMCollectionTypeSet ` as this was the only use case for it